feat: adding break points to `Agent`

davidsbatista opened the pull request:
### Related Issues

- fixes [9532](https://github.com/deepset-ai/haystack/issues/9532), [9534](https://github.com/deepset-ai/haystack/issues/9534) , [9535](https://github.com/deepset-ai/haystack/issues/9535) 

### Proposed Changes:

- Adding breakpoints for Agents using the new datastructure

### How did you test it?

- unit tests, integration tests, manual verification

### Checklist

- I have read the [contributors guidelines](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md) and the [code of conduct](https://github.com/deepset-ai/haystack/blob/main/code_of_conduct.txt)
- I have updated the related issue with new insights and changes
- I added unit tests and updated the docstrings
- I've used one of the [conventional commit types](https://www.conventionalcommits.org/en/v1.0.0/) for my PR title: `fix:`, `feat:`, `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`.
- I documented my code
- I ran [pre-commit hooks](https://github.com/deepset-ai/haystack/blob/main/CONTRIBUTING.md#installation) and fixed any issue

sjrl commented on haystack_experimental/components/agents/agent.py:
```diff
@@ -238,10 +392,41 @@ def run(
+        # resume state if provided
+        if resume_state:
+            component_visits = resume_state.get("pipeline_state", {}).get("component_visits", {})
+            state_data = resume_state.get("pipeline_state", {}).get("inputs", {}).get("state", {}).get("data", {})
```
I think `state_data` need to be deserialized. I believe @Amnah199 can help with this.

davidsbatista commented on haystack_experimental/core/pipeline/pipeline.py:
```diff
@@ -114,9 +116,11 @@ def run(  # noqa: PLR0915, PLR0912
+        :param breakpoints:
+            A set of breakpoints that can be used to debug the pipeline execution.
+
+        :param break_on_first:
+            If `True`, the pipeline will stop at the first breakpoint.
```
exactly that 👍🏽

it was one solution I've found for multiple breakpoints - up until now, we only handled the case for one single breakpoint

Amnah199 replied:
Just thinking out loud.
lets say I want to stop at 3rd breakpoint I can only do that by manual interruption. 

Also if I want to break at the first breakpoint, why would I pass a list of breakpoints 🤔

davidsbatista replied:
Why create an issue? I would rather just leave it as is until there's a clear direction on where to take this new feature.

julian-risch commented on haystack_experimental/components/agents/agent.py:
```diff
@@ -146,6 +152,23 @@ def __init__(
+            if tool_breakpoint.tool_name is not None and tool_breakpoint.tool_name not in available_tool_names:
+                raise ValueError(f"Tool '{tool_breakpoint.tool_name}' is not available in the agent's tools")
+
+        return agent_breakpoint
```
similar to _validate_breakpoint in haystack_experimental/core/pipeline/breakpoint.py

davidsbatista commented:
I forgot to update Haystack to the latest release locally, and I thought the async executor test was still failing.

The integration tests just need a quick fix. Because I did not run them locally, I missed that they were failing.

davidsbatista commented:
I still want to add one more test, but feel free to give it a second round of review.

The main change from the previous revision is that I reverted everything back to a single breakpoint - no more `break_on_first` and the issue with the async is now fixed.

davidsbatista commented:
Note: the re-raise of PBE if Agent raises needs to be done on `pipeline.py` code from main haystack repo - I would rather open an issue and deal with it later or alternatively I can copy the code from the other repo to this one and updated here.

Amnah199 commented on haystack_experimental/core/pipeline/breakpoint.py:
```diff
@@ -7,37 +7,64 @@
+    if isinstance(break_point, Breakpoint) and break_point.component_name not in graph.nodes:
+        raise ValueError(f"pipeline_breakpoint {break_point} is not a registered component in the pipeline")
+
+    if isinstance(break_point, AgentBreakpoint):
+        agent_components = [node for node in graph.nodes if graph.nodes[node]["instance"].__class__.__name__ == "Agent"]
+        if not agent_components:
+            raise ValueError("Defining an AgentBreakpoint but no Agent component found in the pipeline")
+
+        agent_names = [graph.nodes[node]["instance"].__component_name__ for node in agent_components]
+        if break_point.agent_name and break_point.agent_name not in agent_names:
+            raise ValueError(
+                f"pipeline_breakpoint {break_point} is not a registered Agent component in the pipeline. "
+                f"Available agents: {agent_names}"
+            )
+
+        # check if the tool breakpoint is for an actual tool registered in the agent
+        if break_point.tool_breakpoint:
+            agent_component = next(
+                (
+                    node
+                    for node in agent_components
+                    if graph.nodes[node]["instance"].__component_name__ == break_point.agent_name
+                ),
+                None,
+            )
+
+            # get all the tools registered in the agent component
+            if agent_component:
+                agent_instance = graph.nodes[agent_component]["instance"]
+                tools = agent_instance.tools if hasattr(agent_instance, "tools") else []
+                tool_names = [tool.name for tool in tools]
+                if break_point.tool_breakpoint.tool_name and break_point.tool_breakpoint.tool_name not in tool_names:
+                    raise ValueError(
+                        f"pipeline_breakpoint {break_point.tool_breakpoint} is not a registered tool in the agent "
+                        f"{break_point.agent_name}. Available tools: {tool_names}"
+                    )
```
I was wondering if we could simplify this by using something like:

```suggestion
        if isinstance(break_point, AgentBreakpoint):
        breakpoint_agent_component = next(
            (
                node
                for node in graph.nodes
                if graph.nodes[node]["instance"].__component_name__ == break_point.agent_name
            ),
            None,
        )
        if not breakpoint_agent_component:
            raise ValueError(f"pipeline_breakpoint {break_point} is not a registered Agent component in the pipeline")

        if break_point.tool_breakpoint:
            print (graph.nodes[breakpoint_agent_component]["instance"].tools)
            for tool in graph.nodes[breakpoint_agent_component]["instance"].tools:
                if break_point.tool_breakpoint.tool_name == tool.name:
                    break
            else:
                raise ValueError(
                    f"pipeline_breakpoint {break_point.tool_breakpoint} is not a registered tool in the Agent component"
                )   
```

Amnah199 commented:
@davidsbatista I fixed and pushed the commit, looks okay.

sjrl commented on haystack_experimental/dataclasses/breakpoints.py:
```diff
@@ -69,60 +69,26 @@ def __repr__(self):
-                raise ValueError("Breakpoints must be either Breakpoint or ToolBreakpoint.")
+    def __init__(self, agent_name: str, break_point: Union[Breakpoint, ToolBreakpoint]):
+        if not isinstance(break_point, ToolBreakpoint) and break_point.component_name != "chat_generator":
+            raise ValueError("A Breakpoint must have component_name 'chat_generator'.")
```
```suggestion
            raise ValueError("The generator_breakpoint must be a Breakpoint that has the component_name 'chat_generator'.")
```

sjrl commented on haystack_experimental/dataclasses/breakpoints.py:
```diff
@@ -69,60 +69,26 @@ def __repr__(self):
+            raise ValueError("A Breakpoint must be provided.")
 
-        if isinstance(break_point, Breakpoint):
-            if break_point in self.generator_breakpoints:
-                raise ValueError(f"Breakpoint {break_point} already exists in generator breakpoints.")
-            self.generator_breakpoints.add(break_point)
+        self.agent_name = agent_name
+        self.tool_breakpoint = None
+        self.generator_breakpoint = None
```
Just a general question. If we want to restrict users to only be able to pass one breakpoint to AgentBreakpoint, do you think it would make sense to update the dataclass to be something like
```python
class AgentBreakpoint:
    """
    A dataclass to hold a breakpoint that can be used to debug an Agent.
    """
    breakpoint: Union[Breakpoint, ToolBreakPoint]
    agent_name: str = ""
```
I think this would help simplify the `__init__` logic.

davidsbatista replied:
agree 👍🏽 - I will update the code accordingly

davidsbatista commented on haystack_experimental/dataclasses/breakpoints.py:
```diff
@@ -69,60 +69,26 @@ def __repr__(self):
-                self.tool_breakpoints.add(break_point)
-            else:
-                raise ValueError("Breakpoints must be either Breakpoint or ToolBreakpoint.")
+    def __init__(self, agent_name: str, break_point: Union[Breakpoint, ToolBreakpoint]):
```
I think we should instead rename elsewhere to `break_point` since `breakpoint` is a python reserved word:

ruff immediately complains: `A002 Function argument `breakpoint` is shadowing a Python builtin`

Introduced in python 3.7: https://peps.python.org/pep-0553/

sjrl replied:
Sounds good let's use `break_point`

sjrl commented:
Hey @davidsbatista, this is looking great overall! The core logic makes sense to me. There are a few spots where we might be able to simplify and reduce some duplication, but I'm happy to help tackle that in a follow-up after this PR is merged.
